### PR TITLE
Establish TS pattern for core DS components

### LINKED ISSUE #862: Rewrite Box component with TypeScript to establish pattern for core components written in TS

### PR DESCRIPTION
**Pull request by craigpalermo**
This PR configures the tooling required to support TS core components. It also includes a TS refactor of `Box`, `ThemeProvider`, and a few utility functions. The declarations in `src/@types` will also be useful as we begin to refactor more components.

Resolves #862

### CODE REVIEW

File: packages/core/src/Box/Box.tsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  ${space} ${width} ${textAlign} ${boxShadow as never}
+  ${color}
+  ${applyVariations('Box') as never}
+` as React.FunctionComponent<BoxProps>
+
+/**
+ * We can keep the regular prop types around for deprecation warnings at runtime
+ */
+Box.propTypes = {
+  align: deprecatedPropType('textAlign') as never,
```
Comment (craigpalermo) [inline packages/core/src/Box/Box.tsx:47]: TODO: find type for propTypes validators

File: packages/core/src/Box/Box.tsx [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  style?: React.CSSProperties
+}
+
+/**
+ * A low-level layout component for setting width, margin, padding, and color
+ */
+export const Box = styled.div<BoxProps>`
+  ${space} ${width} ${textAlign} ${boxShadow as never}
+  ${color}
+  ${applyVariations('Box') as never}
```
Comment (craigpalermo) [inline packages/core/src/Box/Box.tsx:40]: TODO: find type for styled component style functions

File: packages/core/package.json [hunk 3]
```diff
@@ -9,9 +9,9 @@
   "scripts": {
     "prepare": "run-s clean build",
     "build": "run-s build:js build:types",
-    "build:babel:cjs": "cross-env NODE_ENV=production BABEL_ENV=cjs babel ./src --ignore '**/*.stories.js,**/*.spec.js' --out-dir ./dist/cjs",
-    "build:babel:esm": "cross-env NODE_ENV=production BABEL_ENV=esm babel ./src --ignore '**/*.stories.js,**/*.spec.js' --out-dir ./dist/esm",
-    "build:babel:es": "cross-env NODE_ENV=production BABEL_ENV=es babel ./src --ignore '**/*.stories.js,**/*.spec.js' --out-dir ./dist/es",
+    "build:babel:cjs": "cross-env NODE_ENV=production BABEL_ENV=cjs babel ./src --out-dir ./dist/cjs --extensions \".js,.ts,.tsx\"",
```
Comment (sdalonzo) [inline packages/core/package.json:12]: please don't remove the `--ignore` flags

File: packages/core/src/Flag/__snapshots__/Flag.spec.js.snap [hunk 4]
```diff
@@ -58,7 +58,7 @@ exports[`Flag renders 1`] = `
   display: flex;
 }
 
-@media screen and (min-width:32em) {
+@media screen and (min-width:32) {
```
Comment (sdalonzo) [inline packages/core/src/Flag/__snapshots__/Flag.spec.js.snap:61]: this is a breaking change that needs correction

File: packages/core/src/FormField/__snapshots__/FormField.spec.js.snap [hunk 5]
```diff
@@ -72,7 +72,7 @@ exports[`FormField renders 1`] = `
   font-weight: 700;
 }
 
-@media screen and (min-width:40em) {
+@media screen and (min-width:40) {
```
Comment (sdalonzo) [inline packages/core/src/FormField/__snapshots__/FormField.spec.js.snap:75]: same here

File: packages/core/src/Hide/__snapshots__/Hide.spec.js.snap [hunk 6]
```diff
@@ -7,7 +7,7 @@ exports[`Hide renders 1`] = `
 `;
 
 exports[`Hide renders with lg prop 1`] = `
-@media screen and (min-width:48em) and (max-width:63.99em) {
+@media screen and (min-width:48) and (max-width:63.99em) {
```
Comment (sdalonzo) [inline packages/core/src/Hide/__snapshots__/Hide.spec.js.snap:10]: here as well

File: packages/core/src/IconField/__snapshots__/IconField.spec.js.snap [hunk 7]
```diff
@@ -66,7 +66,7 @@ exports[`IconField renders 1`] = `
   box-shadow: 0 0 0 2px #007aff;
 }
 
-@media screen and (min-width:40em) {
+@media screen and (min-width:40) {
```
Comment (sdalonzo) [inline packages/core/src/IconField/__snapshots__/IconField.spec.js.snap:69]: same

File: packages/core/src/Input/__snapshots__/Input.spec.js.snap [hunk 8]
```diff
@@ -48,7 +48,7 @@ exports[`Input it renders 1`] = `
   box-shadow: 0 0 0 2px #007aff;
 }
 
-@media screen and (min-width:40em) {
+@media screen and (min-width:40) {
```
Comment (sdalonzo) [inline packages/core/src/Input/__snapshots__/Input.spec.js.snap:51]: here as well

File: packages/core/src/Select/__snapshots__/Select.spec.js.snap [hunk 9]
```diff
@@ -58,7 +58,7 @@ exports[`Select renders 1`] = `
   display: none;
 }
 
-@media screen and (min-width:40em) {
+@media screen and (min-width:40) {
```
Comment (sdalonzo) [inline packages/core/src/Select/__snapshots__/Select.spec.js.snap:61]: one more time

File: packages/core/src/theme/__snapshots__/theme.spec.ts.snap [hunk 10]
```diff
@@ -10,11 +10,11 @@ Object {
     "0 0 2px 0 rgba(0,0,0,.08),0 8px 32px 0 rgba(0,0,0,.16)",
   ],
   "breakpoints": Array [
-    "32em",
-    "40em",
-    "48em",
-    "64em",
-    "80em",
+    "32",
```
Comment (sdalonzo) [inline packages/core/src/theme/__snapshots__/theme.spec.ts.snap:13]: same here

File: packages/core/src/theme/theme.ts [hunk 11]
```diff
@@ -1,18 +1,22 @@
...
       enumerable: false,
       get() {
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-return,@typescript-eslint/no-unsafe-member-access
         return this[i]
       },
     })
   )
 
-export const breakpoints = [32, 40, 48, 64, 80].map((n) => n + 'em')
+export const breakpoints = [32, 40, 48, 64, 80].map((n) => `${n}`)
```
Comment (sdalonzo) [inline packages/core/src/theme/theme.ts:19]: I think this is the root cause
```suggestion
export const breakpoints = [32, 40, 48, 64, 80].map((n) => `${n}em`)
```

File: packages/core/src/utils/utils.tsx [hunk 12]
```diff
@@ -170,20 +187,29 @@ export const applyVariations = (componentName, variations = null) => (
...
+
+  // eslint-disable-next-line @typescript-eslint/prefer-regexp-exec
   const colorShade = shade.match(/^([a-z]+)\.([a-z]+)$/i)
 
   if (colorShade) {
     color = colorShade[0]
     shade = colorShade[1]
   }
 
+  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
```
Comment (sdalonzo) [inline packages/core/src/utils/utils.tsx:206]: why are these necessary?

File: packages/core/package.json (hunk 3 above)
Comment (craigpalermo) [reply packages/core/package.json:12]: I moved it to the config so we could keep the scripts a bit shorter https://github.com/priceline/design-system/pull/871/files#diff-fc10a6682269c1fefc2abe05d1500b8aR31

File: packages/core/src/theme/__snapshots__/theme.spec.ts.snap (hunk 10 above)
Comment (craigpalermo) [reply packages/core/src/theme/__snapshots__/theme.spec.ts.snap:13]: I totally missed this, I'll have to look into what's causing it

File: packages/core/src/utils/utils.tsx (hunk 12 above)
Comment (craigpalermo) [reply packages/core/src/utils/utils.tsx:206]: The `themeGet` function from `styled-components` has return type `any`, and TS doesn't like that we're immediately invoking the result of those calls.

**Comment by craigpalermo** [comment]
@sdalonzo Regarding your question about how intellisense will treat DS components written in TS when they're used in JS files, I confirmed that in VSCode and WebStorm that all the prop types defined in the interface appear as expected without having to supplement that with regular `PropTypes`.

Screen Shot 2020-07-08 at 4 52 53 PM

File: packages/core/package.json (hunk 3 above)
Comment (sdalonzo) [reply packages/core/package.json:12]: that works, I missed that. Just verify that they don't appear in `dist`

**Comment by sdalonzo** [comment]
@craigpalermo can you fix the missing coverage?

**Comment by craigpalermo** [comment]
> @craigpalermo can you fix the missing coverage?

I'm not really sure what it's complaining about since all I did was add type info:
Screen Shot 2020-07-09 at 1 25 52 PM

Any ideas?
